subsys/net/lib/azure_fota/azure_fota.c

```diff
@@ -750,3 +766,50 @@ int azure_fota_msg_process(const char *const buf, size_t size)
 
 	return 1;
 }
+
+void azure_fota_modem_firmware_update_validate(void)
+{
+	/* Handle return values relating to modem firmware update */
+	switch (modem_lib_init_result) {
+	case 0:
+		/* Initialization successful, no action required. */
+		return;
+	case MODEM_DFU_RESULT_OK:
+		LOG_DBG("MODEM UPDATE OK. Will run new modem firmware after reboot");
+		break;
+	case MODEM_DFU_RESULT_UUID_ERROR:
+	case MODEM_DFU_RESULT_AUTH_ERROR:
+		LOG_ERR("MODEM UPDATE ERROR %d. Will run old firmware", modem_lib_init_result);
+		break;
+	case MODEM_DFU_RESULT_HARDWARE_ERROR:
+	case MODEM_DFU_RESULT_INTERNAL_ERROR:
+		LOG_ERR("MODEM UPDATE FATAL ERROR %d. Modem failure", modem_lib_init_result);
+		break;
+	default:
+		/* All non-zero return codes other than DFU result codes are
+		 * considered irrecoverable and a reboot is needed.
+		 */
+		LOG_ERR("nRF modem lib initialization failed, error: %d", modem_lib_init_result);
+		break;
+	}
+
+	/* If the Carrier library is enabled, the reboot will be handled there. */
+	if (!IS_ENABLED(CONFIG_LWM2M_CARRIER)) {
+		LOG_DBG("Rebooting...");
+		LOG_PANIC();
+		sys_reboot(SYS_REBOOT_COLD);
+	}
```

Comment:
I feel a  bit weird about 

```
if (!IS_ENABLED(CONFIG_LWM2M_CARRIER)) {
		LOG_DBG("Rebooting...");
		LOG_PANIC();
		sys_reboot(SYS_REBOOT_COLD);
	}
```

specifically. Is it documented anywhere that application developers should avoid rebooting after modem FOTA if LWM2M is enabled? The fact that this is needed is an oddly specific interaction between three areas of our codebase that could easily be missed by an application developer

Reply:
I agree this is a little bit ackward but a nessecity in order to make sure that the carrier library gets to handle the modem return value before the application is rebooted.

@lucasvdp Just to be sure. If the application reboots _before_ the carrier library get to check the return value. Will the FOTA job be marked as success upon the next reboot? Or will the carrier library treat this as an error?